src/pal/src/sharedmemory/sharedmemory.cpp

```diff
@@ -0,0 +1,1114 @@
+        }
+    }
+}
+
+int SharedMemoryHelpers::CreateOrOpenFile(LPCSTR path, bool createIfNotExist, bool *createdRef)
+{
+    _ASSERTE(path != nullptr);
+    _ASSERTE(path[0] != '\0');
+
+    // Try to open the file
+    int openFlags = O_RDWR;
+    while (true)
+    {
+        int fileDescriptor = InternalOpen(path, openFlags);
+        if (fileDescriptor != -1)
+        {
+            if (createdRef != nullptr)
+            {
+                *createdRef = false;
+            }
+            return fileDescriptor;
+        }
+        switch (errno)
+        {
+            case EINTR:
+                continue;
+
+            case ENOENT:
+                if (!createIfNotExist)
+                {
+                    if (createdRef != nullptr)
+                    {
+                        *createdRef = false;
+                    }
+                    return -1;
+                }
+                break;
+
+            case ENAMETOOLONG:
+                throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::NameTooLong));
+
+            case EMFILE:
+            case ENFILE:
+            case ENOMEM:
+                throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::OutOfMemory));
+
+            default:
+                throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+        }
+        break;
+    }
+
+    // File does not exist, create the file
+    openFlags |= O_CREAT | O_EXCL;
+    while (true)
+    {
+        int fileDescriptor = InternalOpen(path, openFlags, PermissionsMask_AllUsers_ReadWrite);
+        if (fileDescriptor != -1)
+        {
+            // The permissions mask passed to open() is filtered by the process' permissions umask, so open() may not set all of
+            // the requested permissions. Use chmod() to set the proper permissions.
+            if (chmod(path, PermissionsMask_AllUsers_ReadWrite) == 0)
+            {
+                if (createdRef != nullptr)
+                {
+                    *createdRef = true;
+                }
+                return fileDescriptor;
+            }
+            CloseFile(fileDescriptor);
+            unlink(path);
+            throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+        }
+        switch (errno)
+        {
+            case EINTR:
+                continue;
+
+            case ENAMETOOLONG:
+                throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::NameTooLong));
+
+            case EMFILE:
+            case ENFILE:
+            case ENOMEM:
+                throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::OutOfMemory));
+
+            default:
+                throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+        }
+    }
+}
+
+void SharedMemoryHelpers::CloseFile(int fileDescriptor)
+{
+    _ASSERTE(fileDescriptor != -1);
+
+    int closeResult;
+    do
+    {
+        closeResult = close(fileDescriptor);
```

Comment:
'close' doesn't return the error code, just -1 if the operation has failed. You need to use errno here.